Mathlib/Combinatorics/Quiver/ConnectedComponent.lean

```diff
@@ -50,18 +57,149 @@ instance [Inhabited V] : Inhabited (WeaklyConnectedComponent V) :=
+    (h : IsSStronglyConnected V) (i j : V) : ∃ p : Path i j, 0 < p.length := h i j
+
+lemma IsSStronglyConnected.exists_pos_cycle {V} [Quiver V]
+    (h : IsSStronglyConnected V) (i : V) : ∃ p : Path i i, 0 < p.length := h i i
+
+lemma IsSStronglyConnected.isStronglyConnected {V} [Quiver V]
+    (h : IsSStronglyConnected V) : IsStronglyConnected V := by
+  intro i j; obtain ⟨p, _⟩ := h i j; exact ⟨p⟩
+
+/-- Equivalence relation identifying vertices connected by directed paths in both directions. -/
+def stronglyConnectedSetoid (V : Type*) [Quiver V] : Setoid V :=
+  ⟨fun a b => (Nonempty (Path a b)) ∧ (Nonempty (Path b a)),
+   fun _ => ⟨⟨Path.nil⟩, ⟨Path.nil⟩⟩,
+   fun ⟨hab, hba⟩ => ⟨hba, hab⟩,
+   fun ⟨hab, hba⟩ ⟨hbc, hcb⟩ =>
+     ⟨⟨hab.some.comp hbc.some⟩, ⟨hcb.some.comp hba.some⟩⟩⟩
+
+/-- The type of strongly connected components (bidirectional reachability classes). -/
+def StronglyConnectedComponent (V : Type*) [Quiver V] : Type _ :=
+  Quotient (stronglyConnectedSetoid V)
+
+namespace StronglyConnectedComponent
+
+variable {V}
+
+/-- The canonical map from a vertex to its strongly connected component. -/
+protected def mk : V → StronglyConnectedComponent V :=
+  @Quotient.mk' _ (stronglyConnectedSetoid V)
+
+instance : Coe V (StronglyConnectedComponent V) :=
+  ⟨StronglyConnectedComponent.mk⟩
+
+instance [Inhabited V] : Inhabited (StronglyConnectedComponent V) :=
+  ⟨(default : V)⟩
+
+protected lemma eq (a b : V) :
+  (a : StronglyConnectedComponent V) = b
+    ↔ (Nonempty (Path a b) ∧ Nonempty (Path b a)) := Quotient.eq''
+
+@[simp] lemma mk_eq_mk {a b : V} :
+    (StronglyConnectedComponent.mk a : StronglyConnectedComponent V) =
+    StronglyConnectedComponent.mk b ↔
+    (Nonempty (Path a b) ∧ Nonempty (Path b a)) :=
+  StronglyConnectedComponent.eq a b
+
+lemma IsSStronglyConnected.pos_cycle
+    {V} [Quiver V] (h : IsSStronglyConnected V) (v : V) :
+    ∃ p : Path v v, 0 < p.length := h v v
+
+end StronglyConnectedComponent
+
+variable {V}
+
+lemma stronglyConnectedComponent_eq_of_path {a b : V}
+    (hab : Nonempty (Path a b)) (hba : Nonempty (Path b a)) :
+    (a : StronglyConnectedComponent V) = b :=
+  (StronglyConnectedComponent.eq (a := a) (b := b)).2 ⟨hab, hba⟩
+
+lemma exists_path_of_stronglyConnectedComponent_eq {a b : V}
+    (h : (a : StronglyConnectedComponent V) = b) :
+    (Nonempty (Path a b)) ∧ (Nonempty (Path b a)) :=
+  (StronglyConnectedComponent.eq (a := a) (b := b)).1 h
+
+lemma stronglyConnectedComponent_singleton_iff {V} [Quiver V] (v : V) :
+    (∀ w : V, (w : StronglyConnectedComponent V) = v → w = v) ↔
+    (∀ w : V, w ≠ v → ¬(Nonempty (Path v w) ∧ Nonempty (Path w v))) := by
+  constructor
+  · intro h_singleton w hw_ne h_bidir
+    obtain ⟨hab, hba⟩ := h_bidir
+    have h_same_scc : (w : StronglyConnectedComponent V) = v :=
+      stronglyConnectedComponent_eq_of_path (a := w) (b := v) hba hab
+    obtain rfl := h_singleton w h_same_scc
+    exact hw_ne rfl
```

Comment:
This is a total nitpick but do get the magical `rfl` desruction (which rewrites equalities automatically) you need to do:
```suggestion
    obtain ⟨rfl⟩ := h_singleton w h_same_scc
    contradiction
```